### DISCUSSION TITLE
Q: Is type checking done by semantic analyzer?

### DISCUSSION THREAD

**Discussion by Glyphack** [pgrandinetti/compilers#4] [Q&A] [answered]
I'm pretty new to the semantic analysis concept. In the guide we are calling the binding + type checking semantic analysis. But I'm seeing different names for semantic analysis step in different projects.

For example pyright has separate components for this binding type:
https://github.com/microsoft/pyright/blob/fb56dccde02874f2db2c0bbe64fa76fc469fd690/docs/internals.md#analysis-phases

And mypy calls the binding phase semantic analysis:
https://github.com/python/mypy/wiki/Semantic-Analyzer

I'm curios to know if there's any different between semantic analysis, and binding + type checking.

**Comment by pgrandinetti** [comment]
Thanks for the links. I need some time to look at those references. Just to give a quick answer in the meantime, my references are the dragon book and the compilers class at stanford.

**Comment by pgrandinetti** [comment, marked as answer]
According to my notes: yes, type checking is (typically) done in the Semantic Analysis step.
Broadly speaking:
1. Lexer: divides the input in tokens.
2. Parser: transforms the sequence of tokens in the parse tree.
3. Semantic analyzer: perform several operations, including type checking, error reporting and it tries to identify as many consistencies as possible (which often depends on the language specifications at hand).
4. Optimizer: perform smart transformation of the parse tree to improve the code.
5. Code generator: transforms the optimized parse tree in machine code.

Ref. Dragon book 2nd edition, Section 1.2 (specifically, 1.2.3)

> I'm curios to know if there's any different between semantic analysis, and binding + type checking.

Again according to my notes and references, binding+type checking is just one of the things done by a Semantic Analyzer.
